### ISSUE TITLE
Specifying output names in v4.x when running parallel jobs on cluster

### ISSUE THREAD

**Issue by charlesdavid** [biocore/sortmerna#216]
Dear developers, I have been using your software since version 2 and want to point out some potential loss of functionality. Please understand that this feedback is not meant to be critical but to alert you to some of the issues that your users are encountering and to support your efforts to improve your program.
- In version 2.1b, the user could specify the output paths of the aligned and other reads which made using sortmerna in workflows convenient. That ability is not present in version 4.0.0. 
- When one considers running a workflow to filter rRNA with 100's or 1000's of samples, it does not make sense to calculate the indices for the 8 references for every sample, which is now the case (due to the structure of the native workdir). 
  - Also, pipelines that use your program for rRNA filtering become much more difficult to implement. 

- The biggest issues with version 2.1b were the need to merge and de-merge files for paired end reads and the need for uncompressed files. You have solved most of that problem... you just need to output the F and R reads in two files. 
  - This is important for your software as most pipelines use paired end data. 
    - **I suggest that you bring back the original options that allow the user to direct the output according to their needs.** 
  - Also, when running many samples it would be convenient to have an option to point to a static directory of indices.  
    - The `-workdir` flag should remain so that a temp directory for the key-value work can be created for each sample, SEPARATE FROM the output and indices.

There may also be some performance issues in all versions:
- We have processed merged fastqs of size 50GB with version 2.1b and the processing time is quite long, running for about 4 days to search the 8 databases. 
- We have noticed that no matter how many cores are specified, the program only creates 3 PIDs and uses only 3 threads. This occurs for versions 3 and 4 as well. 
  - Note, we are using your pre-compiled binaries on a Centos 6 HPC.
- While testing versions 3 and 4, we noticed that even after 1 day, there is NO OUTPUT at all. This is very different from version 2 and quite confusing as no errors were thrown to either stdout or stderr.

I will continue to provide you with feedback from our use of your software and want to encourage you to keep striving for improvement as your program addresses a specific need in the bioinformatics community. Thanks for all your efforts.

**Comment by biocodz** [comment]
@charlesdavid 
Hi, thank you for the feedback. We welcome a constructive critics:-)

The many changes and bug fixes introduced since version 2.x changed the way users interact with the program, which is always an inconvenience. Many problems the users report stem from an incorrect use. We'll do our best to satisfy the user requests in this respect.

The coming version 4.1.0 has option to output paired reads into separate files.

(!) If you don't see a living (progressing) execution trace like this [Sample trace](https://github.com/biocore/sortmerna/wiki/Sample-execution-trace), Don't Wait for days - stop immediately, and log an issue.

There is No Need to repeat indexing while changing the samples. If the `workdir` containing indices is reused, the indices already present in `workdir/idx/` Are Reused.

If you want to preserve the output of a previous run, just rename the `workdir/out/` to e.g. `workir/out_20200218_718/`

The only directory that has to be removed prior each run is `workdir/kvdb/` for the time being. See [here](https://github.com/biocore/sortmerna/wiki/2.-User-manual-(todo)#Work_directory)

As for multithreading, by default the program creates as many processing threads as there are cores on your host. You can adjust the thread count using `-a` option. Please, open a separate issue if you have a problem.

Please, open issues for each problem separately other than bundling them all together.

**Comment by charlesdavid** [comment]
We have workflows for many samples... OK, so I have 1000 samples and specify a single  `--workdir /a/dir/ ` that has /a/dir/idx for all those 1000 samples.
- Now, are all output files for those 1000 samples going to this directory: /wkdir/out? 
- If so, how will they be named?
  - Since I can not specify a name, I guess they will all have the same name... so a big mess.
- It seems that I need to create 1000 separate directories and copy the indices 1000 times...
  - Then the output is distributed over 1000 directories but all with the same names...

- How do I get my output files named according to their proper sample name and whether it is filtered? 
  - **What I really need is to direct the output as in 2.1b**. 
- How do I do that with 4.0.1?
- The challenge in software development is NOT to break the existing good features while adding new ones!

Thanks.

**Comment by elias22b** [comment]
> We have workflows for many samples... OK, so I have 1000 samples and specify a single `--workdir /a/dir/ ` that has /a/dir/idx for all those 1000 samples.
> 
> * Now, are all output files for those 1000 samples going to this directory: /wkdir/out?
> * If so, how will they be named?
>   
>   * Since I can not specify a name, I guess they will all have the same name... so a big mess.
> * It seems that I need to create 1000 separate directories and copy the indices 1000 times...
>   
>   * Then the output is distributed over 1000 directories but all with the same names...
> * How do I get my output files named according to their proper sample name and whether it is filtered?
>   
>   * **What I really need is to direct the output as in 2.1b**.
> * How do I do that with 4.0.1?
> * The challenge in software development is NOT to break the existing good features while adding new ones!
> 
> Thanks.

I am having the same issue when running multiple jobs (i.e. separate directory for each sample, and all output files for the samples have the same name). Currently I find it's more practical to gunzip, interleave, run 2.1b, deinterleave, and gzip.

Hopefully in the future it will be possible to specifically name the aligned.fastq and other.fastq files with sortmerna 4.

**Comment by biocodz** [comment]
Does 1000 samples mean you have 1000 reads files? If so, you'd like to specify 1000 different names for the output files. Correct?

**Comment by elias22b** [comment]
> Does 1000 samples mean you have 1000 reads files? If so, you'd like to specify 1000 different names for the output files. Correct?

Considering I've a similar issue I will also reply. I do not have 1000 samples, but about 50 samples sequenced on NovaSeq S4 (~90 million sequences per sample).
This means 50 R1 files, and 50 R2 files.
When running sortmerna4 I will start one job for each sample (i.e. 50 jobs) and each job will have a specific --workdir created to prevent kvdb to already exist.

**Comment by biocodz** [comment]
From what I understand, you have to run the SMR 50 times. You could do it the following way.
```
sortmerna -ref 1 -ref 2 ... -reads r1_fwd -reads r1_rev -workidr ~/smr
```
This would create 
```
/smr/idx/
/smr/kvdb/
/smr/out/
  aligned.fasta
  other.fasta
  ...
```
After the first run you could
```
mv ~/smr/out ~/smr/out_1/
// and do the second run
rm -rf ~/smr/kvdb/       // do prior each new run
sortmerna -ref 1 -ref 2 ... -reads r2_fwd -reads r2_rev -workidr ~/smr
```
The above will reuse the indices.
Continue in this manner with all the 50 files. You'll end up with
```
~/smr/idx/
~/smr/kvdb/
~/smr/out_1/
~/smr/out_2/
...
~/smr/out_50/
```

**Comment by elias22b** [comment]
> From what I understand, you have to run the SMR 50 times. You could do it the following way.
> 
> ```
> sortmerna -ref 1 -ref 2 ... -reads r1_fwd -reads r1_rev -workidr ~/smr
> ```
> 
> This would create
> 
> ```
> /smr/idx/
> /smr/kvdb/
> /smr/out/
>   aligned.fasta
>   other.fasta
>   ...
> ```
> 
> After the first run you could
> 
> ```
> mv ~/smr/out ~/smr/out_1/
> // and do the second run
> rm -rf ~/smr/kvdb/       // do prior each new run
> sortmerna -ref 1 -ref 2 ... -reads r2_fwd -reads r2_rev -workidr ~/smr
> ```
> 
> The above will reuse the indices.
> Continue in this manner with all the 50 files. You'll end up with
> 
> ```
> ~/smr/idx/
> ~/smr/kvdb/
> ~/smr/out_1/
> ~/smr/out_2/
> ...
> ~/smr/out_50/
> ```

Hi,
it's not a problem to run it sortmerna 50 times. I run them as 50 parallel slurm jobs using a for loop. Each job take several days to finish so they have to run simultaneously or this will tkae a few months to finish.

The cumbersome issue with sortmerna4 is that I haven't found a way to directly specify the name of align.fastq and other.fastq when running sortmerna4 (which was possible with 2.1b). But maybe there is such a command?

**Comment by biocodz** [comment]
I see the problem now. Indeed, starting multiple smr jobs in parallel won't work as the output names are set in stone. Specifying multiple different workdir is a non-efficient solution because of the necessity to duplicate the indices.
We'll add the ability to specify the names for the output directories and/or files in the release 4.1.1. 4.1.0 is currently being tested.

**Comment by elias22b** [comment]
> I see the problem now. Indeed, starting multiple smr jobs in parallel won't work as the output names are set in stone. Specifying multiple different workdir is a non-efficient solution because of the necessity to duplicate the indices.
> We'll add the ability to specify the names for the output directories and/or files in the release 4.1.1. 4.1.0 is currently being tested.

That's great to hear! Good luck and thanks for continuing developing on sortmerna.

**Comment by elias22b** [comment]
> I see the problem now. Indeed, starting multiple smr jobs in parallel won't work as the output names are set in stone. Specifying multiple different workdir is a non-efficient solution because of the necessity to duplicate the indices.
> We'll add the ability to specify the names for the output directories and/or files in the release 4.1.1. 4.1.0 is currently being tested.
> 
> Would the option `-other dirname` `-aligned dirname` be good enough for the start or would you better recommend different file names?

I think a good start would be able to name the output files. For example, -sample_id sample1 with the output files being named sample1_other.fastq and sample1_aligned.fastq.

And for  parallel jobs it will be sample2, sample3 etc.  It then wouldn't matter if I make 50 different --workdirs (or if it will be possible to get all the output files in one workdir directory). I could just move all the files out from these directories and organize them (considering they name have different names I will know what samples belonging to which files).

**Comment by biocodz** [comment]
Actually separate output for paired reads is already implemented, but I'm still not through with the testing. Hopefully soon.

Running a 'parallel' app in parallel on the same machine is a sub-optimal solution. An uncharted territory really. I didn't even keep in mind such a use case - the reason for my slow reaction:-)
I guess we need to make SMR to be able to process multiple butches of reads, so that the program could optimize the usage of the resources.

**Comment by elias22b** [comment]
> Actually separate output for paired reads is already implemented, but I'm still not through with the testing. Hopefully soon.
> 
> Running a 'parallel' app in parallel on the same machine is a sub-optimal solution. An uncharted territory really. I didn't even keep in mind such a use case - the reason for my slow reaction:-)
> I guess we need to make SMR to be able to process multiple butches of reads, so that the program could optimize the usage of the resources.

I am not sure how much of a problem this is. I would guess that most scientists with this kind of data uses a High-performance Computational Cluster provided by national or university infrastructure. So each app/job is run on different nodes in parallel.

**Comment by biocodz** [comment]
I didn't know the cluster was smart enough to figure out sharing the resources between those jobs.

**Comment by biocodz** [comment]
The following arguments will be added in the next release:
```
-aligned   optional    [dir/][pfx]    Default WORKDIR/out/aligned
-other     optional    [dir/][pfx]      dir and file prefix for non-aligned output
-kvdb      optional    [dir]          key-value DB location. Default: WORKDIR/kvdb/
-idx       optional    [dir]          index directory, Default: WORKDIR/idx/
```
Details on `-aligned`:
```
Directory and file prefix to use for aligned output i.e. each output file will go into 
the specified directory with the given prefix. 
The appropriate extension will be automatically added.

If 'dir' is not specified, the output will be created in the provided or the default WORKDIR

If 'pfx' is not specified, the prefix 'aligned' will be used
	   
 Examples:
 -aligned $MYDIR/dir_1/dir_2/1  -> $MYDIR/dir_1/dir_2/1.fasta|fastq|blast|sam
 -aligned dir_1/bla             -> $PWD/dir_1/bla.fasta|sam|blast
 -aligned bla                   -> WORKDIR/bla.fasta|fastq|sam|blast
 -aligned  (no argument)        -> WORKDIR/aligned.fasta/fastq/sam/blast 
```
`-other` is similar to `-aligned`

Note that existing since v2.x `-pid` option can always be used to generate unique file names per a run e.g.
```
sortmerna -ref .. -read .. -pid

produces:

aligned_17536.fastq
aligned_17536.log
other_17536.fastq
```

**Comment by elias22b** [comment]
Hi,
I've now tested the new changes in 4.1 and 4.2 and everything so far works great!

**Comment by biocodz** [comment]
Hi,
Good news! Thank a lot for mentioning this.
biocodz

**Comment by biocodz** [comment]
resolved in v4.2.0

**Comment by Ibberson-Lab** [comment]
I was able to run multiple samples using version 4.0.0 Created a for loop that could run sortmerna, rename the out directory of each sequence and purge the kvdb directory of each run before proceeding to the next and I got desired results.
